### ISSUE TITLE
[Feature Request] Adding Cobra as a control mechanism

### ISSUE THREAD

**Issue by vaugenwake** [Melkeydev/go-blueprint#157]
### Tell us about your feature request

Many projects need more than just an API. For example, most projects that use a DB would also need to implement a way to migrate a database schema or start a background worker task to consumer jobs, etc.

I propose an optimisation to the project structure, which is to add Cobra as a control interface for the project so that you can execute these operations separately and scale them horizontally later on if needed.

Example:
```BASH
make run start <-- starts the API server
make run migrate <-- execute migrations with something like Goose or Gorm
make run consumer <-- maybe some background worker
```

The proposal here is not to add these features but to abstract starting the API into a cobra command and make cobra a dependency of the generated project structure.

P.S. I am quite new to Go and learning it, and this has been a great starting point. Just an observation from other workflows for example Laravel

Would be happy to work on implementing this if it is something that is wanted.

### Disclaimer

- [X] I agree

**Comment by Ujstor** [comment]
A DB migration feature would be nice. Recently, I created app with go-blueprint and htmx. For DB migration, I had to use a Docker workaround. If you are interested, this is yours. Ensure that DB migration is only available when users are using DB drivers

**Comment by Melkeydev** [comment]
I would be super interested in this

**Comment by vaugenwake** [comment]
This was my first attempted implementation of migrations using Goose
https://github.com/vaugenwake/posts-api

Think goose might be a good generic implementation not needing to use any ORM or framework and allows users who would like to keep using RAW SQL to do so as the default.

Also changed a few other things like running the complete Go appl in docker, this is not required but would make local networking simpilar.

**README has details about how to run the project.**

Would this be an approach that would be ideal? Keen for your thoughts @Ujstor @Melkeydev

**Comment by vaugenwake** [comment]
Should probably split the two topics of Migrations and adding Cobra?

**Comment by Ujstor** [comment]
Nice concept with a well executed implementation. While there were some small bugs (initial migration doesn't work), this  is still a work in progress. If in the template injection level is executed correctly, it has the potential to be perfect. Personally, I am always in favor of using Docker. 

[DB management tool](https://www.adminer.org/) is nice touch, handy.  There is PR #151 with default .env variables, and if this gets implemented, there is a need to change DB_HOST to belong to DNS resolution from the Docker Host.

Also, maybe we can use API endpoints with the current htmx implementation for demonstrative purposes, writing and retrieving data from the database. @briancbarrow  This is just an idea, what do you think?

On the other hand, things are getting complicated very quickly, and I have fears that we are pushing the blueprint in a direction where it becomes too customized. This feature can be instantly added as an advanced flag, the question is whether it should be included in the base blueprint

I am interested in what thoughts others have about this. Personally, I would like to see a migrations tool implemented.

For the Cobra control mechanism, I am not sure how that would look. You can do a simple implementation that will give more insight into the functionality. In any case, you should split these two into separate parts

Nice job

**Comment by Ujstor** [comment]
![Screenshot from 2023-12-28 20-59-26](https://github.com/Melkeydev/go-blueprint/assets/116409846/8b5cf0aa-57f6-4a14-8468-9d338f9d1e77)

![Screenshot from 2023-12-28 20-58-33](https://github.com/Melkeydev/go-blueprint/assets/116409846/000282e5-f86b-495d-864d-b8b782998351)

![Screenshot from 2023-12-28 20-58-51](https://github.com/Melkeydev/go-blueprint/assets/116409846/4b537052-d220-4ee6-8b0f-a853c93c967f)

![Screenshot from 2023-12-28 20-46-26](https://github.com/Melkeydev/go-blueprint/assets/116409846/e6a883be-6888-4183-b97f-1fd2537d9797)

**Comment by vaugenwake** [comment]
@Ujstor Here is an example of Cobra being implemented on a blueprint project: https://github.com/vaugenwake/cobra-blueprint

Doing this allows you to run other concerns your application might have. Items like queue consumers, cron jobs, migrations etc can be setup like this and controlled independently or scaled independently.

Example your API might need/want 2 containers running for redudency and scale up on burst traffic, but your queue consumer might only need a single container and scale on message volume, but both of these concerns share logic between them makes no sense to copy logic or deploy 2 applications or to scale your consumers because your API has to scale out.

**Comment by Ujstor** [comment]
Sorry for delay...For now, we would like to see the implementation of Goose in the base blueprint, excluding containers. You can utilize a Makefile implementation similar to Air. The goal is to maintain simplicity and enable users to easily modify created project for their specific use cases.

You can introduce a fully dockerize app using the advanced flag. Concept you provided is good.

For Cobra, we are not sure; this is a significant change. We are not against it, we would like to see how it is implemented.

From my standpoint, you can create three separate PR-s, with the Cobra PR scheduled for the end.

**Comment by vaugenwake** [comment]
@Ujstor Fair enough, will look at a simple addition for Goose when a compatible DB driver is selected. Will reach out if I need anything.

**Comment by vaugenwake** [comment]
Still working on this, will try and get something in before the end of Jan

**Comment by Ujstor** [comment]
DB migration (Goose) is still needed, but due to recent blueprint development direction, we believe that Cobra impl is not suitable for the project.
